src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -391,8 +392,451 @@ internal static JsonNodeType ReadNext(this IJsonReader jsonReader)
+        internal static Task ReadStartObjectAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            return ReadNextAsync(jsonReader, JsonNodeType.StartObject);
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is an EndObject node.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+		/// <returns>A task that represents the asynchronous read operation.</returns>
+        internal static Task ReadEndObjectAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            return ReadNextAsync(jsonReader, JsonNodeType.EndObject);
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is an StartArray node.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+		/// <returns>A task that represents the asynchronous read operation.</returns>
+        internal static Task ReadStartArrayAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            return ReadNextAsync(jsonReader, JsonNodeType.StartArray);
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is an EndArray node.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+		/// <returns>A task that represents the asynchronous read operation.</returns>
+        internal static Task ReadEndArrayAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            return ReadNextAsync(jsonReader, JsonNodeType.EndArray);
+        }
+
+        /// <summary>
+        /// Verifies that the current node is a property node and returns the property name.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+		/// The value of the TResult parameter contains the property name of the current property node.</returns>
+        internal static Task<string> GetPropertyNameAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+            Debug.Assert(jsonReader.NodeType == JsonNodeType.Property, "jsonReader.NodeType == JsonNodeType.Property");
+
+            // NOTE: the JSON reader already verifies that property names are strings and not null/empty
+            return jsonReader.GetValueAsync().FollowOnSuccessWith(tsk => (string)tsk.Result);

```

Comment:
@habbes The equivalent of what we're doing here is making the method async and then changing the return statement to this:
```csharp
object value = await jsonReader.GetValueAsync().ConfigureAwait(false);
return (string)value;
```
We're only making the method async so we await the call to `jsonReader.GetValueAsync()` and then cast the value into a string.
The idea behind using `FollowOnSuccessWith` is to avoid making the method async (and avoid the generation of the state machine). There are overheads from the use of `FollowOnSuccessWith` method too but I'll do a few tests to accurately determine which approach is better from a COGS perspective.

Reply:
@habbes Behind the scenes it uses `ContinueWith`. That logic predates async/await. For now I have replaced with the async/await equivalent